### DISCUSSION TITLE
Add sensor device classes for apparent and reactive power

### DISCUSSION THREAD

**Discussion by dmulcahey** [home-assistant/architecture#688] [Entity Models] [answered]
## Decision:
Add new sensor device classes "apparent power" (allowed unit `VA`) and "reactive power" (allowed unit `var`)

Core PR: https://github.com/home-assistant/core/pull/62010

## Original suggestion:
Currently we only support W and kW for sensors with a device class of power:

https://developers.home-assistant.io/docs/core/entity/sensor/#available-device-classes

 Should we add additional units of measurement to support additional devices? See this issue for context: 

https://github.com/home-assistant/core/issues/57236

**Comment by DrCoolzic** [comment]
I am obviously bias but I think it is important for HA to support true (P), apparent (S) and reactive (Q) power expressed respectively in Watts, VA, and VAR
For technical detail please refer (for example) to [True, Reactive, and Apparent Power](https://www.allaboutcircuits.com/textbook/alternating-current/chpt-11/true-reactive-and-apparent-power/)

**Comment by emontnemery** [comment, marked as answer]
Power is measured in W, apparent and reactive power is not, and it's not possible to convert either reactive power or apparent power to power without additional information.

We should hence add new device classes "apparent power" (allowed unit `VA`) and "reactive power" (allowed unit `var`) instead.

**Reply by toreamun** [reply]
What about kVAr? Other related power units like watt support both W and kW.
kVAr is a common native unit from electrical household power meters. 

Reactive energy unit varh/kvarh is units that goes hand in hand with reactive power (var/kvar) in the same way as energy (wh/kwh) and power (W/kw). Electrical household power meters that measure reactive power usually also measure reactive energy. At least this is the case in the national standards in Norway and Sweden. With reactive power device class in place, a reactive energy device class has become the missing piece in the puzzle.

**Reply by DrCoolzic** [reply]
For information the French Linky energy meter also return varh

**Reply by emontnemery** [reply]
This discussion already has an accepted answer, and the new device classes and units were added by https://github.com/home-assistant/core/pull/62010

I suggest to open a new discussion suggesting to add additional device class and units for reactive energy.

**Reply by toreamun** [reply]
Thanks. I then added #724

**Comment by MartinHjelmare** [comment]
Can we estimate how many integrations can leverage the new device classes?

**Reply by dmulcahey** [reply]
I can list several:

ZHA
Deconz
Z2m
maybe ZwaveJS? Zwave devices tend to be similar to Zigbee but you guys would know better
esphome - from the linked issue

Potentially other integrations like SmartThings that bridge their stuff into HA as well.

**Reply by tomsoftcom** [reply]
tasmota via mqtt as well

**Reply by dmulcahey** [reply]
Hrm wonder if capturing a relationship between integration to things like this (device classes, uom, etc?) could be useful for the statistics that are in the analytics portal… not necessarily to display but to aid in situations like this?

**Reply by farmio** [reply]
`Fronius` would too, as would probably any other solar inverter integration or power-meter reading integration. 

And of course `knx` and `modbus` as they allow devices_class definition via yaml - but both do have plenty devices which would use it - especially Modbus is widely used directly in power meters and solar inverters (at least in Europe).

**Reply by emontnemery** [reply]
Native Tasmota can use this too.

**Reply by MartinHjelmare** [reply]
Here are the ones I found that use the concerned units explicitly:

VA:
apucpsd
fronius
growatt_server
iotawatt
mysensors
nut
tasmota
zha

var:
fronius
iotawatt
mysensors

**Reply by MartinHjelmare** [reply]
Considering the above, the support for a device class for apparent power seems enough. For reactive power the support is not as strong, as three integrations is on the low side. Can we find some more?

**Reply by emontnemery** [reply]
For `var`:
- `tasmota`: The library (HATasmota) has support, but it's not enabled in the integration due to the lack of support in Home Assistant
- `dsmr`: Reactive power is reported by some meters (Swedish, maybe others), but this data is not used in the integration

**Reply by MartinHjelmare** [reply]
Good! Five is enough, I think.

**Reply by DrCoolzic** [reply]
FYI many ESPHome energy monitoring modules return VAR (as well as VA) for example ATM90E32, SDM Energy monitor, Select Meter, Teleinfo and probably more ...
PZEM AC does not directly return VAR but it can be computed easily with lambda as PF is provided

**Reply by farmio** [reply]
And in case we need some backups, again 
- `knx` and 
- `modbus` 

support `var`
And they also support the energy pendants `VAh` and `varh` (so does `Fronius`)

**Reply by DrCoolzic** [reply]
I just checked the French power meter specifications and to my surprise it does provide varh (but not vah) to measure reactive energy reinjected in the grid for people with solar panel. But this is rather esoteric and I am not sure anybody will use this information

**Comment by DrCoolzic** [comment]
~~Is there really a need for 3 powers?~~

~~- The "Real Power" or "True Power» or "Active Power"~~
~~- The "apparent power"~~
~~- The "reactive power"~~

In grid distribution the most important power is the “apparent power” because this is what unfortunately causes lots of drops in voltage, needs for bigger transformer, for thicker wires, etc. To a point that some power providers are thinking of using of apparent power to charge customers (actually this is already the case in France for some large industries …). So yes, apparent power is a critical information.

~~If it is easier to create three device classes why not. But personally, I believe that allowing the three units of measurement (W, VA, VAR) in the power class is just fine unless we want to be academically perfect~~

**Reply by MartinHjelmare** [reply]
Each device class should only have one unit of measurement with room for internationalization.

**Reply by DrCoolzic** [reply]
~~What about pressure: cbar, bar, hPa, inHg, kPa, mbar, Pa, psi
temperature~~

**Reply by dgomes** [reply]
How many households are concerned with apparent power and reactive power ? These are very important for industrial consumers, but we are **Home** Assistant.

**Reply by emontnemery** [reply]
@DrCoolzic "apparent power" and "reactive power" is not the same thing as "power", and we can't convert between them.
Pressure or temperature expressed in different units are all measuring the same physical quantity, with units other than SI units used in some regions or for certain applications - that's the internationalization Martin is referring to.

@dgomes yes, but there are consumer grade power meters which do report `VA` or `var`, including Swedish smart electricity meters 🤷‍♂️

**Reply by DrCoolzic** [reply]
Not that it is interesting but as said apparent power is provided because it is easy to get s=va
for real power you need integral computation so often not provided or at least not real time. for example in the French Linky meter the real power is computed every  30 mn.
![image](https://user-images.githubusercontent.com/6078364/146034517-7aa6ea52-a0dc-4d40-a5fe-a154bd3c663e.png)

At the end you are mainly only interested by energy but **real time** power consumption even if approximate is nice to follow
![image](https://user-images.githubusercontent.com/6078364/146033603-aa786778-9968-4ed1-b1a0-3901b23bc408.png)

computing real power and power factor is a bit more tricky and delayed
![image](https://user-images.githubusercontent.com/6078364/146033836-cab52986-2b0f-4c81-8155-a46f1a3de0be.png)

"Apparent power" is an interesting value anyway because it is always bigger than real power so it gives you an enveloppe. 
It is also good to know that usually AT HOME (not in the industry) when the "apparent power" raise a lot it is usually due to resistive load (like radiators, heaters...) and in that case the power factor gets close to one i.e. P get close to Q. 
At the opposite when power is very low (the noise level) then the power  factor degrade due to electronic component with bad PF but in that case as the power is very low you dont really care :)

**Reply by DrCoolzic** [reply]
> @DrCoolzic "apparent power" and "reactive power" is not the same thing as "power", and we can't convert between them. Pressure or temperature expressed in different units are all measuring the same physical quantity, with units other than SI units used in some regions or for certain applications - that's the internationalization Martin is referring to.
> 
Sorry you are right
> @dgomes yes, but there are consumer grade power meters which do report `VA` or `var`, including Swedish smart electricity meters 🤷‍♂️

French and Belgium too. In the French Linky apparent power is the only real time power indicator. As mentioned Real power can only be obtained by integral computation and is a discrete value delayed by 30 mn and only available every 30 mn. So only useful for historical reference **but not for triggering real time automation**

**Comment by DrCoolzic** [comment]
Any schedule for this enhancement? Meanwhile HA reports error and refuses to record my power !

**Reply by MartinHjelmare** [reply]
Apparent and reactive power sensor device classes were added half a year ago:
https://github.com/home-assistant/core/pull/62010

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

It's up to each integration to leverage them.

**Comment by DrCoolzic** [comment]
Now I am really confused. Maybe I did not explain correctly? it all started here https://github.com/home-assistant/core/issues/57236
I have some ESPHome sensors that reports apparent power in VA and this works well. The problem is that in statistics this unit is reported as an error:
![image](https://user-images.githubusercontent.com/6078364/175541149-cc298b4e-96be-4f9b-af0d-a691d4efc638.png)

And this is still the case apparently the power is NOT recorded in the “statistic database” and therefore we cannot use for example the “statics graph card” to display result.

**Reply by MartinHjelmare** [reply]
Issues with integrations are handled in the core repo. Please open a new issue there if there's still a problem. 👍

**Comment by DrCoolzic** [comment]
Sorry to bother you but I do not understand what you mean by issue with the integration. I need some to help to understand where is the problem and I would appreciate if you can help.
I have created an ESHHome device that returns many sensor. For example one of them is the apparent power computed by the electricity meter. The sensor looks like this
![image](https://user-images.githubusercontent.com/6078364/175647811-196bf47a-7ec3-494a-9c61-52d0dc536f5c.png)
long long time ago when statistic where first introduced this sensor was recorded in the statistic db. But some more checking where added and since then this sensor is no more recorded as statistic and therefore not available to display.
I just recompiled with latest ESPHome version but still not working ? Whhat integration needs to be fixed ? ESPHome ? recorder ? ???
Thanks in advance if you can tell me where I need to place a request or if I am doing something wrong

**Reply by parautenbach** [reply]
> I do not understand what you mean by issue with the integration

It means this is an architectural discussion thread (the future), not the core system's issue tracker (bugs) and not the forum (support).  In other words, log the issue against the HA core repo, but perhaps best to seek help and confirm the issue on the forum first.

**Comment by DrCoolzic** [comment]
oops I just realized that I should probably use "apparent power" and not "power"
so stupid
sorry

**Comment by DrCoolzic** [comment]
Just tested but unfortunately Esphome does not support it but now I know where to report
